File: examples/languages/javascript/orderService.js
export class OrderService {
  constructor(catalog, repository) {
    this.catalog = catalog;
    this.repository = repository;
  }

  create(productId, quantity) {
    const product = this.catalog.find(productId);
    if (!product) {
      throw new Error(`Unknown product: ${productId}`);
    }

    const order = {
      productId,
      quantity,
      total: product.price * quantity,
    };
    this.repository.save(order);
    return order;
  }
}

File: examples/languages/kotlin/OrderService.kt
data class Product(val price: Int)

data class Order(
    val productId: String,
    val quantity: Int,
    val total: Int,
)

interface Catalog {
    fun find(productId: String): Product?
}

interface OrderRepository {
    fun save(order: Order)
}

class OrderService(private val catalog: Catalog, private val repository: OrderRepository) {
    fun create(productId: String, quantity: Int): Order {
        val product = catalog.find(productId)
        if (product == null) {
            error("Unknown product: $productId")
        }

        val order = Order(productId, quantity, product.price * quantity)
        repository.save(order)
        return order
    }
}
